File: Hash Table/888_Fair-Candy-Swap/888_Fair-Candy-Swap.cpp
/*******************************************************************************
 * Title:
 *     888. Fair Candy Swap
 *     888. 公平的糖果棒交换
 * Address: 
 *     https://leetcode-cn.com/problems/fair-candy-swap/
 ******************************************************************************/

#include <iostream>
#include <vector>
#include <numeric>
#include <set>
using namespace std;

// 方法一：哈希表
class Solution {
public:
    vector<int> fairCandySwap(vector<int>& A, vector<int>& B) {
        vector<int> ans;

        long sumA = accumulate( A.begin(), A.end(), 0 );
        long sumB = accumulate( B.begin(), B.end(), 0 );
        int b = ( sumB - sumA ) / 2;  // y = x + b;

        set<int> hashSet( B.begin(), B.end() );

        for ( int a : A ) {
            if ( hashSet.count( a + b ) ) {
                ans.push_back( a );
                ans.push_back( a + b );

                break;
            }
        }

        return ans;
    }
};


int main()
{
    

    return 0;
}

File: Hash Table/888_Fair-Candy-Swap/888_Fair-Candy-Swap.py
"""
 * Title:
 *     888. Fair Candy Swap
 *     888. 公平的糖果棒交换
 * Address: 
 *     https://leetcode-cn.com/problems/fair-candy-swap/
"""

class Solution:
    def fairCandySwap(self, A: List[int], B: List[int]) -> List[int]:
        sumA, sumB = sum(A), sum(B)
        delta = (sumB - sumA) // 2  # sumB 与 sumA 相减的先后顺序，会影响到后续的 for 循环

        rec = set(B)
        ans = list()

        for a in A:
            b = a + delta
            if b in rec:
                ans = [a, b]
                break
        
        return ans
                